hangtime: implement a leaky bucket algorithm

mlyle opened the pull request:
This allows power added to exceed the configured value for a short
time-- to prevent tumbling when chopping throttle from
fast-forward-flight (where aerodynamic loads require high stabilization
power) and to allow entry to aggressive maneuvers while in hangtime
regimes.

mlyle committed:
hangtime: implement a leaky bucket algorithm

This allows power added to exceed the configured value for a short
time-- to prevent tumbling when chopping throttle from
fast-forward-flight (where aerodynamic loads require high stabilization
power) and to allow entry to aggressive maneuvers while in hangtime
regimes.

mlyle commented:
This is -critical- code, so I need to review this myself first before going any further.

mlyle commented:
Here's how it looks if you're at 90% throttle, chop, and control loops want to use 100% authority:

![image](https://user-images.githubusercontent.com/90903/29841269-a85044b6-8cb9-11e7-8236-c5cebd1ba524.png)

mlyle commented:
Here's how it looks if you are at 90% throttle, chop, and the stabilization loop wants 30% power, steady state.

![image](https://user-images.githubusercontent.com/90903/29841379-1ade0680-8cba-11e7-8d25-b05621990e1a.png)

mlyle commented:
Here's what it looks like if you chop from 90% power, and the control loop wants to use 60% power, 50ms on, 50ms off.

![image](https://user-images.githubusercontent.com/90903/29841486-6f0487f2-8cba-11e7-939c-80a211b8fe1f.png)

mlyle commented:
X axis, all graphs: time.  y axis, all graphs: power added.  Power is chopped / demands start being made at 2s.  Maxpoweradd configured as 15%.

mlyle commented:
One last plot, probably the most realistic:

![image](https://user-images.githubusercontent.com/90903/29841716-61117960-8cbb-11e7-9ead-32049315b29e.png)

90% power chopped;control loop wants to use 45% authority (2 motors at 90%) 450ms on, 150ms off.  Seems to have a pretty sane result for what the control loop is actually granted.

glowtape commented:
I'm curious as to how the chopped power method flies and the effect it has on the motors.

mlyle commented:
They're not different methods.. The 'chopped' are showing what happens to delivered power when there are intermittent stabilization power demands (control deflections, etc)

mlyle commented:
Just for clarity--- old way: hangtime allows up to 15% power for stabilization, which means it can e.g. if maxpower add is set to 15%, spin 2 motors up to 30%.  But the stabilization power demands are intermittent, so the average is guaranteed to be less than 15% and you can't do any maneuver that takes more power than this.

The new way: the long term **average** is guaranteed to be 15% or less, and the maximum instantaneous power used after some time in hangtime is 30% (double 15%).  But for a short time the limits are much looser, depending on how much power you were using before.

If you're in fast-forward-flight and the rear motors are at 45% more power than the fronts to stop from tipping over, the old way would reduce in the thing aerodynamically tipping over if you went to hangtime and not that much power was available.  Now there's a little bit of time when the limits are much more lax, so hopefully the vehicle can slow down and aerodynamically unload in that time.

mlyle committed:
refactor hangtime time limit to stabilization

It was previously in actuator, which is a bit of a nasty place for it.
In turn, it's complicated the path to 3D stab, and also does not use
I term stabilization when in hangtime.

Note existing configurations will be lost-- as the field is in a
different uavo now (StabilizationSettings instead of ActuatorSettings).

mlyle commented:
Further extended to do the hangtime decision making up in stabilization so that the integral term can still do integrally stuff.

This requires careful verification.  I have already read diff and exercised some in simulator.

tracernz commented on flight/Modules/Actuator/actuator.c:
```diff
@@ -311,15 +311,62 @@ static void fill_desired_vector(
+	 * with time; at high throttle it corresponds to 300ms of the current
+	 * power; at lower throttle it corresponds to 300ms of double the
+	 * configured value.
+	 */
```
👍

tracernz reviewed (approved):
Looks good. 👍

glowtape commented:
Just went for a test flight. During hangtime, my quad maintains attitude nicely, now that the I-term works. When cutting quickly, you see a tiny jolt in the goggles, but the I-term seems to catch up in a fraction of a second and things seem stable.

However, the time limit doesn't kick in. I can put my quad on the ground, power it up, and when I arm, the motors spin idly. A quick throttle jolt and back to zero throttle (so it sits on the ground), hangtime kicks in, as expected, but doesn't time out on the 4.5s I've configured. It keeps slowly spooling up on the gyro noise, until it eventually wants to flip. If you disarming and rearming, it just continues. Until I power cycle.

glowtape commented:
On the bench, after hangtime, throttle only goes to 1,00001e-37 according to GCS, not true zero. Might be that.

glowtape commented:
Nevermind, it's my ManualControl interpolator mode. I have to fix that.

mlyle committed:
stabilization: introduce THROTTLE_EPSILON as threshold

mlyle commented:
Rebased to be on top of #1952 for safety reasons.

mlyle commented:
Ready for merge, I think.
